### ISSUE TITLE
Learning meaningful location embeddings from unlabeled visits

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1319]
Learning meaningful location embeddings from unlabeled visits<br>
<br>
https://ift.tt/2OYytS5<br>
<br>
<br>
<br>
<div><div data-title="Learning meaningful location embeddings from unlabeled visits" data-url="https://www.sentiance.com/?p=9170">
<div><div><div><div><div><div>
<h2><span>Introduction</span></h2>
<p>At Sentiance, we developed an AI platform that learns to detect and predict a person’s <a rel="noopener" href="http://www.sentiance.com/2016/05/31/how-ai-drives-the-mobile-contextual-revolution/" target="_blank">behavior, routines and profile</a> from mobile sensor data such as accelerometer, gyroscope and GPS. Many low level algorithms, such as our <a rel="noopener" href="http://www.sentiance.com/2017/12/04/large-scale-transport-mode-classification-from-mobile-sensor-data/" target="_blank">transport mode detector</a>, are trained by supervised learning techniques for which we gathered labeled data by closely collaborating with our partners, customers and test groups. However, supervised learning comes with at least two major drawbacks:</p>
<ol>
<li>Gathering labeled data in a natural setting can be difficult and very expensive.</li>
<li>The algorithm does not improve over time, even when millions of unlabeled data points per day are captured.</li>
</ol>
<p>A concrete example that would benefit from unsupervised or semi-supervised learning, is venue mapping, i.e. figuring out which venue a user is visiting given a noisy location estimate. Since the type of place a user is currently visiting highly depends on his recent and long-term past behavior, this kind of data needs to be gathered in a realistic setting where the temporal behavior of the user actually corresponds to natural human behavior. While gathering labeled data for venue mapping is a tedious process, getting unlabeled data comes for free as more users are using our platform. The question then becomes: Can we extract meaningful information from population-wide human visiting patterns without requiring access to the actual labels?</p>
<p>The main motivation for this project is therefore to enable us to learn from these large unlabeled datasets, with minimal preprocessing and feature engineering. More specifically, we want to learn a representation of a location that embeds different types of meaningful information, for example the category (e.g. shop, school, etc.) of the place present at that location, and some information about population-wide behavior at that location. We can then use these representations in our venue mapping and other models.</p>
</div></div></div></div></div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Hypothesis-compressor-1.png" width="13670" height="7110"></div>
</figure>
</div>
<div><div><div><div><div><div>
<p>To achieve this, we adapt the DeepCity deep learning framework <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>, which is inspired on word2vec word embeddings <i>(<a rel="noopener" href="https://arxiv.org/abs/1301.3781" target="_blank">Mikolov, 2013</a>) </i>and graph embeddings<i> (<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>)</i>.</p>
<h2><span>DeepCity</span></h2>
<p>DeepCity<i> (<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i> is a feature learning framework based on deep learning, to profile users and locations. It uses check-in data obtained from online social networks such as Facebook, Twitter, Foursquare, etc. Users frequently share their whereabouts using those platforms in the form of so-called check-ins. Two use cases are considered: user profiling, and location profiling.  In the context of user profiling, it is assumed that the whereabouts of a user reflects who he or she is. User profiling includes tasks such as age or gender prediction and is the focus of Pang et al. Location profiling, on the other hand, is aimed at learning more about locations and includes tasks such as predicting the category of a location. The latter is the focus of this blog, because of the direct link to venue mapping.</p>
<p>User and location profiling tasks are being handled by machine learning algorithms, typically involving hand-engineered features. The latter is very time consuming and requires domain-specific knowledge. Even with expert domain knowledge, it is difficult to capture all relevant features, that are applicable in all scenarios. Deep learning provides a promising alternative, where features are no longer hand engineered but are learned by the model itself. DeepCity <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i> can be considered such an approach, specifically for user and location profiling.</p>
<p>DeepCity uses a graph representation of check-ins and random walks on this graph and outputs node embeddings. More specifically, location and user nodes are organized in a bipartite graph, and task-specific random walks are used to explore the graph. Using those random walks and some clever algorithms we can learn our embeddings. But let’s take a step back first, and put everything into context.</p>
</div></div></div></div></div></div>
<div><div><div><div>
<div><div>
<h2><span>From word embeddings to graph embeddings</span></h2>
<p>DeepCity <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i> is based on graph or network embeddings <i>(<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>)</i> using the skip-gram model <i>(<a rel="noopener" href="https://arxiv.org/abs/1301.3781" target="_blank">Mikolov, 2013</a>)</i>. The objective is to learn representations (embeddings) that preserve each user’s or location’s neighbor information. We then use these embeddings as features for other models (e.g., user or location profiling). In addition, we can fine tune these representations to specific tasks. Manual engineered features are no longer needed.</p>
<p>To understand the type of graph embeddings proposed in the DeepCity paper we first need to go back to natural language processing, more specifically the word2vec and the skip-gram model <i>(<a rel="noopener" href="https://arxiv.org/abs/1301.3781" target="_blank">Mikolov, 2013</a>)</i>. The goal of the word2vec algorithm is to learn a representation for each word given a corpus. It became known and popular due to interesting properties, such as the ability to do arithmetics with them. A famous example:</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Figure-1@10x-compressor-300x216.png" width="300" height="216"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div><div><div>
<p><b><i>Figure 1</i></b><i>:</i> <i> <a rel="noopener" href="https://blog.acolyer.org/2016/04/21/the-amazing-power-of-word-vectors/" target="_blank">Word vector examples for ‘king’, ‘man’, ‘queen’, and ‘woman’.</a> </i></p>
<p>The word2vec algorithm has also been adapted to graphs, for example in the context of social networks. DeepWalk <i>(<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>),</i> among others, uses word2vec in the context of graphs: graph nodes correspond to words, and random walks on the graph are used to generate the sentences. Below is an example from the DeepWalk paper <i>(<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>)</i>. The input (left) is the famous Zachary’s karate club graph where nodes are assigned a label (color), and the outputs (right) are the node embeddings. The embeddings capture the structure and labels of the network. More recently related approaches have started to show up in the specific context of location based social networks <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>.</p>
</div></div></div></div></div></div>
<div><div><div><div><div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/deepwalk1-2.png" width="946" height="308"></div>
</figure>
</div></div></div></div></div>
<div><div><div><div><div><div>
<p><b>Figure 2:</b> (left) <a href="https://en.wikipedia.org/wiki/Zachary%27s_karate_club">Zachary’s Karate Club Graph</a>, and (right) DeepWalk representation. Image Source: DeepWalk paper (<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>)</p>
<h2><span>Graph representation of check-ins</span></h2>
<h3><span>Graph construction</span></h3>
<p>Similar to the DeepWalk algorithm <i>(<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>)</i> described above, DeepCity <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i> uses a graph representation and random walks, and outputs node embeddings. More specifically, locations and users are organised in a bipartite graph based, and (task-specific) random walks are used to explore the graph.</p>
<p>The input to the DeepCity algorithm consists of <i>(timestamp, user id, venue id)</i> check-in records. From those check-ins, we can construct a bipartite graph, that links users and the venues they visit. Let’s consider a small example check-in dataset. The dataset consists of two users that work in the same office. One of the users typically gets lunch at a nearby supermarket, whereas the other user often brings lunch but sometimes gets a sandwich.</p>
</div></div></div></div></div></div>
<div><div><div><div><table><tbody>
<tr>
<th>User</th>
<th>Venue</th>
<th>Check-ins</th>
<th>Description</th>
</tr>
<tr>
<td>u<sub>1</sub>
</td>
<td>v<sub>1</sub>
</td>
<td>5</td>
<td>Supermarket</td>
</tr>
<tr>
<td>u<sub>1</sub>
</td>
<td>v<sub>2</sub>
</td>
<td>5</td>
<td>Office</td>
</tr>
<tr>
<td>u<sub>2</sub>
</td>
<td>v<sub>2</sub>
</td>
<td>4</td>
<td>Office</td>
</tr>
<tr>
<td>u<sub>2</sub>
</td>
<td>v<sub>3</sub>
</td>
<td>1</td>
<td>Sandwich Shop</td>
</tr>
</tbody></table></div></div></div></div>
<div><div><div><div>
<div><p>We can represent the check-in dataset as a bipartite graph, linking users and visited venues.</p></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Figure3-compressor-300x269.png" width="300" height="269"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div><div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Figure4-compressor.png" width="14520" height="4640"></div>
</figure>
</div></div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 4:</b> (left) Example (unweighted) bipartite check-in graph, (center) weighted check-in graph (users → venues), (right) weighted check-in graph (venues → users).</p>
<h3><span>Random walks</span></h3>
<p>Now that we have a graph representation of the visits, where the graph nodes can be considered words in word2vec sense. We now need a way to generate sentences from this graph representation. For this, we use random walks. At each node in the random walk, we randomly sample one of the neighbors based on the edge weight. Below is an example random walk sequence.</p>
</div></div>
<table><tbody>
<tr>
<th>Step</th>
<th>Random Walk Sequence</th>
</tr>
<tr>
<td>0</td>
<td>u<sub>1</sub> → v<sub>1</sub>
</td>
</tr>
<tr>
<td>1</td>
<td>u<sub>1</sub> → v<sub>1</sub> → u<sub>1</sub>
</td>
</tr>
<tr>
<td>2</td>
<td>u<sub>1</sub> → v<sub>1</sub> → u<sub>1</sub> → v<sub>2</sub>
</td>
</tr>
<tr>
<td>3</td>
<td>u<sub>1</sub> → v<sub>1</sub> → u<sub>1</sub> → v<sub>2</sub> → u<sub>2</sub>
</td>
</tr>
<tr>
<td>4</td>
<td>u<sub>1</sub> → v<sub>1</sub> → u<sub>1</sub> → v<sub>2</sub> → u<sub>2</sub> → v<sub>3</sub>
</td>
</tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p>We can now assume that the graph nodes (temporal users and venues) correspond to words, and that the random walks correspond to sentences. We can now apply the word2vec algorithm (<a rel="noopener" href="https://arxiv.org/abs/1301.3781" target="_blank">Mikolov, 2013</a>) to the random walk sequences and obtain embeddings of the graph nodes.  However, the authors of the DeepCity paper actually go one step further and propose task-specific (or biased) random walks, which we will consider next.</p>
<p>The DeepCity framework <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i> uses task-specific random walks, which use location and user properties to guide the feature learning to be more specific to each task. In other words, the random walks are guided towards certain users or locations which are more relevant to the prediction task in mind.</p>
<p>To achieve this, a measurement to quantify to which extent a user is biased towards location categories is proposed. We can determine the check-in distribution of the population over the location categories, and for each user. A user is assumed to be biased when his or her check-in category distribution <b>P</b> is different from the population check-in category distribution <b>Q</b>. The authors propose the Kullback-Leibler divergence to define such a bias: <b>KL(Q|| P).</b> As a result, users who exhibit behavior similar to that of the whole population get low bias values, whereas users with different behavior get higher bias values.</p>
<p>Assume a category bias value <img src="https://www.sentiance.com/wp-content/ql-cache/quicklatex.com-8e1c480f44a4ea406d0f066c11c9e2c0_l3.svg" alt="x_{u, category}" width="74" title="Rendered by QuickLaTeX.com" height="14"> for user <img src="https://www.sentiance.com/wp-content/ql-cache/quicklatex.com-43fe27dc3e528266a619764d90fce60b_l3.svg" alt="u" width="10" title="Rendered by QuickLaTeX.com" height="8">, we then obtain the category biased edge weight from venue <img src="https://www.sentiance.com/wp-content/ql-cache/quicklatex.com-ef71511c70f0e4b25cc6bd69f3bc20c2_l3.svg" alt="v" width="9" title="Rendered by QuickLaTeX.com" height="8"> to user <img src="https://www.sentiance.com/wp-content/ql-cache/quicklatex.com-43fe27dc3e528266a619764d90fce60b_l3.svg" alt="u" width="10" title="Rendered by QuickLaTeX.com" height="8">:</p>
<p><img src="https://www.sentiance.com/wp-content/ql-cache/quicklatex.com-1e1975b8d853b049090a866cb03dbc81_l3.svg" alt="w^{category}_{v,u} = w_{v, u} , x_{u, {category}}" width="199" title="Rendered by QuickLaTeX.com" height="24"></p>
<p>The edge weight from user to venue remains unchanged. The underlying idea is that we want to visit more interesting users more often in the random walks and see what other places they visit. We show an example below of a check-in category distribution of the population (left) and example user (right).</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/fig_5_population_vs_example.png" width="877" height="338"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div><div><div>
<p><b>Figure 5:</b> (left) Population, (right) Example user</p>
<p>Note that in order to perform task-specific random walks we need labeled data, i.e., we need to be able to associate a category with each venue. If users check-in to specific venues (e.g., using online social networks), this is easily obtained, however when only a (latitude, longitude) pair is available, like in our situation, this is no longer the case and we need some way to approximate this information in order to use task-specific walks. We will discuss such modifications later, but first, we will have a look at some experimental results.</p>
</div></div></div></div></div></div>
<div><div><div><div>
<div><div>
<h2><span>DeepCity for labeled data</span></h2>
<h3><span>Experimental Setup</span></h3>
<p>For the purpose of this blog post we used the TSMC2014 NYC check-in dataset <i>(Yang, 2014) </i>for all research experiments. The dataset consists of about 10 months of check-in data in New-York and was collected from Foursquare between April 2012 and February 2013. Each anonymized user in the dataset has at least 100 check-ins, but a venue may have only a single check-in.</p>
<p>We present the dataset statistics of the raw dataset and those of the filtered dataset used in the experiments. In addition, we also include the statistics of the dataset used in the DeepCity paper <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>, although they can not be compared due to different data collection procedures, filtering, etc. We filtered the NYC dataset <i>(Yang, 2014)</i> by requiring at least 10 check-ins at each venue, which is in line of the experiments in <i>(Pang, 2017)</i>, and by removing certain categories, for example, those with a temporary nature such as events, or without a fixed location.</p>
</div></div>
<table><tbody>
<tr>
<th></th>
<th>Raw</th>
<th>Filtered</th>
<th>DeepCity</th>
</tr>
<tr>
<td><b>Venues</b></td>
<td>38.333</td>
<td>34.937</td>
<td>54.961</td>
</tr>
<tr>
<td><b>Check-ins</b></td>
<td>227.428</td>
<td>194.258</td>
<td>19.600.207</td>
</tr>
<tr>
<td><b>Users</b></td>
<td>1.083</td>
<td>1.083</td>
<td>1.165.286</td>
</tr>
<tr>
<td><b>Active users</b></td>
<td>/</td>
<td>/</td>
<td>12.778</td>
</tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div><div><div>
<p>We trained 128-dimensional embeddings over 50 epochs using a linear decreasing learning rate. During an epoch, a random walk of length 100 is generated for each venue and temporal user node. We use the skip-gram architecture with negative sampling (5 negative samples) and a context window of size 11.</p>
<p>A logistic regression classifier is used for the location profiling task, i.e., given a venue embedding predict the corresponding venue category. We used 5-fold cross-validation and report the average results over the folds. Grid search was used to determine the optimal hyper-parameters (e.g., regularization).</p>
</div></div></div></div></div></div>
<div><div><div><div>
<div><div>
<h3><span>Exploring DeepCity Embeddings</span></h3>
<p>Let’s start by taking a look at the learned embeddings before we start building other models on top of them. Below we show a 2D representation of the embeddings obtained by applying the T-SNE algorithm. We colored each venue node according to its category, and are able to discover structure in it, which we highlighted.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/tsne_annotated.png" width="1288" height="1144"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 6:</b> DeepCity venue embeddings visualized using t-SNE. Venue nodes are colored according to their venue category.</p>
<p>Let’s see how subcategories are organized. We now consider the Travel &amp; Transport category and corresponding subcategories. In the PCA based representations below we filtered on this category, and highlight a specific subcategory.</p>
<p>Notice for example that Hotels (left) are grouped together in one part of the space, and airport-related venues (right) are grouped together in another part of the space.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Figure-7-compressor-1024x451.png" width="640" height="282"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 7:</b> DeepCity travel &amp; transport venue embeddings visualized using PCA. (Left): Travel &amp; Transport (white), Hotels (red). (Right) Travel &amp; Transport (white), Airport (red)</p>
<p>We can also take a look at the different transport modes used in New York, and again we notice that they are grouped together in somewhat distinct areas of the space, e.g., metro in the left figure, bus in the right figure.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Figure-8-compressor-1024x434.png" width="640" height="271"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 8:</b> DeepCity travel &amp; transport venue embeddings visualised using PCA. (Left): Travel &amp; Transport (white), Subway (red). (Right) Travel &amp; Transport (white), Bus (red)</p>
<p>Word embeddings, such as those obtained from the word2vec algorithm described earlier, are based on the distributional hypothesis: <i>“You shall know a word by the company it keeps” (Firth, 1957). </i>A similar idea exists in the context of geography, and is called Tobler’s first law of geography: <i>“Everything is related to everything else, but near things are more related than distant things.” (Tobler, 1969). </i>Let’s explore these ideas a bit further. The venues are this time colored according to their postal code. Notice that the embeddings capture the geospatial relations, i.e., nodes with the same color cluster together.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/Figure-9-compressor-1024x825.png" width="640" height="516"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 9:</b><b>  </b>DeepCity travel &amp; transport venue embeddings visualized using PCA. Venue embeddings colored according to postal code.</p>
<p>These visualizations show that the resulting embeddings capture both geo-spatial distance and semantic similarity at the same time, distributed over different dimensions of the embedding vector.</p>
<p>Given a certain venue as query point, we can also easily determine the nearest neighbors in the embedding space. Let’s determine the five nearest neighbors for some example venues.</p>
</div></div>
<table><tbody>
<tr>
<th>Venue</th>
<th>Category</th>
<th>Distance</th>
</tr>
<tr>
<td>Terminal 3</td>
<td>Travel &amp; Transport</td>
<td>0.460</td>
</tr>
<tr>
<td>JetBlue Crew Lounge</td>
<td>Travel &amp; Transport</td>
<td>0.478</td>
</tr>
<tr>
<td>Jb Shuttle Stop – T5</td>
<td>Travel &amp; Transport</td>
<td>0.486</td>
</tr>
<tr>
<td>Delta Sky Club</td>
<td>Travel &amp; Transport</td>
<td>0.503</td>
</tr>
<tr>
<td>JB Shuttle Stop – Metry &amp; Lefferts</td>
<td>Travel &amp; Transport</td>
<td>0.538</td>
</tr>
<tr><td colspan="3"><b><u>Query:</u> John F. Kennedy International Airport (Travel &amp; Transport)</b></td></tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p>Notice that the query results are all related to travel and transport, even though there are other types of venues located in the same airport, that are also present in our dataset. This again shows that both geography and semantic similarity are capture simultaneously.</p>
<p>Let’s consider another example, but this time related to related arts and entertainment.</p>
</div></div>
<table><tbody>
<tr>
<th>Venue</th>
<th>Category</th>
<th>Distance</th>
</tr>
<tr>
<td>The Metropolitan Area</td>
<td>Arts &amp; Entertainment</td>
<td>0.248</td>
</tr>
<tr>
<td>The Joyce Theater</td>
<td>Arts &amp; Entertainment</td>
<td>0.272</td>
</tr>
<tr>
<td>Elinor Bunin Munroe Film Center</td>
<td>Arts &amp; Entertainment</td>
<td>0.291</td>
</tr>
<tr>
<td>BAM Harvey Theater</td>
<td>Arts &amp; Entertainment</td>
<td>0.317</td>
</tr>
<tr>
<td>Brooklyn Sports Club</td>
<td>Outdoor &amp; Recreation</td>
<td>0.328</td>
</tr>
<tr><td colspan="3"><b><u>Query:</u> Carnegie Hall (Arts &amp; Entertainment)</b></td></tr>
</tbody></table>
<div><p>Again, the the nearest neighbours are also related to arts and entertainment, but there is one somewhat unexpected venue present. Let’s consider a final example related to nightlife.</p></div>
<table><tbody>
<tr>
<th>Venue</th>
<th>Category</th>
<th>Distance</th>
</tr>
<tr>
<td>PH-D at Dream Downtown</td>
<td>Nightlife Spot</td>
<td>0.369</td>
</tr>
<tr>
<td>Mother’s ruin</td>
<td>Nightlife Spot</td>
<td>0.380</td>
</tr>
<tr>
<td>Wharf Bar &amp; Grill</td>
<td>Nightlife Spot</td>
<td>0.408</td>
</tr>
<tr>
<td>Electric Room</td>
<td>Nightlife Spot</td>
<td>0.412</td>
</tr>
<tr>
<td>Top of the Standard</td>
<td>Nightlife Spot</td>
<td>0.417</td>
</tr>
<tr><td colspan="3"><b><u>Query:</u> Le Bain (Nightlife Spot)</b></td></tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div><div><p>From these initial observations, we already get indications that the Deepcity framework embeds meaningful information for location profiling purposes, such as categories, subcategories, and geospatial information.</p></div></div></div></div></div>
<div><div><div><div>
<div><div>
<h3><span>Location Profiling</span></h3>
<p>Let us now consider the location profiling task, our main interest. Given a location embedding, we want to predict the category associated with this location.  Before presenting our results, let us first consider the results presented in <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>, using their custom NYC dataset.</p>
</div></div>
<table><tbody>
<tr>
<th></th>
<th>Micro-F1</th>
<th>Macro-F1</th>
<th>Weighted-F1</th>
</tr>
<tr>
<td>Unbiased</td>
<td>51.00%</td>
<td>33.00%</td>
<td>/</td>
</tr>
<tr>
<td>Biased</td>
<td>63.00%</td>
<td>48.00%</td>
<td>/</td>
</tr>
<tr><td colspan="4"><b>Reference results (Pang, 2017)</b></td></tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p>Below are the results using the TSMC2014 NYC dataset (Yang, 2014). We include the results from a few trivial classifiers to help putting the results in context</p>
<ul>
<li>
<b>Most frequent: </b>always predict the most frequent category in the dataset.</li>
<li>
<strong>Uniform:</strong> sample uniformly from the categories.</li>
<li>
<b>Stratified:</b> sample from the categories based on the distribution present in the dataset.</li>
</ul>
</div></div>
<table><tbody>
<tr>
<th></th>
<th>Micro-F1</th>
<th>Macro-F1</th>
<th>Weighted-F1</th>
</tr>
<tr>
<td>Most Frequent</td>
<td>28.03%</td>
<td>5.47%</td>
<td>12.27%</td>
</tr>
<tr>
<td>Uniform</td>
<td>11.39%</td>
<td>10.54%</td>
<td>12.42%</td>
</tr>
<tr>
<td>Stratified</td>
<td>16.65%</td>
<td>12.40%</td>
<td>16.58%</td>
</tr>
<tr>
<td>Unbiased</td>
<td>38.06%</td>
<td>35.81%</td>
<td>37.56%</td>
</tr>
<tr>
<td>Biased</td>
<td>41.24%</td>
<td>39.09%</td>
<td>41.15%</td>
</tr>
<tr><td colspan="4"><b>Sentiance results</b></td></tr>
</tbody></table>
<div><p>Although we are using a similar type of dataset (i.e., Foursquare check-ins in NYC), our results are not as good as reported in the paper <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>. However, we do observe a similar influence of the type of random walks. Although the results are not ideal (i.e., we would have preferred the same or better scores) they still demonstrate the potential of the method.</p></div>
</div></div></div></div>
<div><div><div><div><div><div>
<h3><span>Sparse Labels</span></h3>
<p>In the DeepWalk paper <i>(<a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Perozzi, 2014</a>) </i>the authors considered the scenario of sparsely labeled training data, i.e., only a limited amount of (node) labels are available. This is interesting because we do have a limited amount of labeled data that we can use. We, therefore, experimented with sparse labels in the context of the DeepCity framework <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>.</p>
<p>We first learned embeddings using the regular (unbiased) random walks and task-specific (biased) random walks. We then trained a classifier that predicts the category given an embedding. We kept the test set fixed, but trained the classifier using increasing amounts of data.</p>
</div></div></div></div></div></div>
<div><div><div><div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/fig_10_sparse_labels-1.png" width="698" height="266"></div>
</figure>
</div>
<div><div>
<p><b>Figure 10:</b><b>  </b>Evaluation of location profiling performance in function of the amount of training data.</p>
<p>We obtain the best results when all available training data is used, but the approach is also promising when only a limited amount of data is available.</p>
</div></div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<h2><span>Extending DeepCity to unlabeled datasets</span></h2>
<h3><span>General approach and initial results</span></h3>
<p>Until now we have considered the DeepCity framework <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i> in the context of (labeled) check-in datasets, where we know the exact venue or place that is being visited. We must adapt the framework in order to support our (unlabeled) stationary datasets. The difference between the two types of data is shown in the table below.</p>
</div></div>
<table><tbody>
<tr>
<th>Type</th>
<th>Data</th>
<th>Data Collection</th>
</tr>
<tr>
<td><b>Stationary</b></td>
<td><ul>
<li>User</li>
<li>Start and end timestamps</li>
<li>Location (latitude, longitude)</li>
</ul></td>
<td>Implicit<br>
(By our SDK)</td>
</tr>
<tr>
<td><b>Check-in</b></td>
<td><ul>
<li>User</li>
<li>Start timestamp</li>
<li>Place ID</li>
</ul></td>
<td>Explicit<br>
(By manual user action)</td>
</tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div><div><div>
<p>We first needed to move from an infinite amount of often noisy (latitude,longitude) pairs, to a discrete set of locations or places. Clustering algorithms may come to mind, but would add additional complexity. What we need is a simple and deterministic way to map (latitude, longitude) coordinates to a discrete set of locations. Geohashes are therefore an ideal candidate, which represent latitude and longitude information as a string, where the precision depends on the number of characters used.</p>
<p>We now have the minimum requirements for the DeepCity framework <i>(<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>)</i>:  check-in tuples (user, timestamp, place). However, at this point we can only use unbiased random walks, which are not tuned to the specific task under consideration (e.g., location profiling). In order to support task specific (biased) random walks, we need contextual data (labels).</p>
<p>We need contextual information for two purposes. First we need contextual data for task-specific walks. Second, we need ground truth contextual information to evaluate the embeddings in the location profiling task, i.e., category prediction. Check-in datasets from location based social networks do include this type of information, but as mentioned before, we do not always have access to such ground truth data. Can we make an informed guess on the category to assign to a certain geohash (area)?</p>
<p>The basic idea is to consult a point-of-interest provider to find nearby venues and pick the most common category among those venues. Clearly, this is a very crude approximation, but it is trivial to implement. Of course, we could use more elaborate approaches, but we intentionally kept things as simple as possible. We consider two point-of-interest providers or views:</p>
<ul>
<li>
<b>Check-in view: </b>We extract venues from the check-in dataself itself. Consequently, we will only be aware of venues for which check-ins were observed during the data collection period. Note that other venues may exist in the area, but those are unobserved. In other words, this view is related to what the users actually do.<br><table><tbody>
<tr>
<th></th>
<th>Micro-F1</th>
<th>Macro-F1</th>
<th>Weighted-F1</th>
</tr>
<tr>
<td>Most Frequent</td>
<td>31.08%</td>
<td>5.93%</td>
<td>14.74%</td>
</tr>
<tr>
<td>Uniform</td>
<td>11.40%</td>
<td>9.94%</td>
<td>13.26%</td>
</tr>
<tr>
<td>Stratified</td>
<td>20.44%</td>
<td>12.19%</td>
<td>20.33%</td>
</tr>
<tr>
<td>Unbiased</td>
<td>33.81%</td>
<td>29.04%</td>
<td>34.87%</td>
</tr>
<tr>
<td>Biased</td>
<td>38.16%</td>
<td>32.14%</td>
<td>38.82%</td>
</tr>
<tr><td colspan="4"><b>Sentiance results: Check-in View</b></td></tr>
</tbody></table>
</li>
<li>
<b>Infrastructure view: </b>We obtain venues directly from the point-of-interest provider, which will contain the unobserved venues that are missing in the check-ins view. In other words, this view is related to which venues are located in the area, which does not necessarily reflect what users actually do.<br><table><tbody>
<tr>
<th></th>
<th>Micro-F1</th>
<th>Macro-F1</th>
<th>Weighted-F1</th>
</tr>
<tr>
<td>Most Frequent</td>
<td>40.41%</td>
<td>7.20%</td>
<td>23.26%</td>
</tr>
<tr>
<td>Uniform</td>
<td>11.69%</td>
<td>9.30%</td>
<td>14.18%</td>
</tr>
<tr>
<td>Stratified</td>
<td>27.27%</td>
<td>12.81%</td>
<td>26.95%</td>
</tr>
<tr>
<td>Unbiased</td>
<td>27.27%</td>
<td>19.74%</td>
<td>29.88%</td>
</tr>
<tr>
<td>Biased</td>
<td>31.47%</td>
<td>21.20%</td>
<td>33.95%</td>
</tr>
<tr><td colspan="4"><b>Sentiance results: Infrastructure View</b></td></tr>
</tbody></table>
</li>
</ul>
</div></div></div></div></div></div>
<div><div><div><div>
<div><div>
<h3><span>Influence of noisy labels</span></h3>
<p>To evaluate the impact of noise introduced by the place and context approximations we consider the following scenarios:</p>
<p>Scenario 1:</p>
<ol>
<li>Train embeddings using the infrastructure view.</li>
<li>Train classifier using the checkins view.</li>
<li>Evaluate classifier using the checkins view.</li>
</ol>
<p>Scenario 2:</p>
<ol>
<li>Train embeddings using the infrastructure view.</li>
<li>Train classifier using the infrastructure view.</li>
<li>Evaluate classifier using the checkins view.</li>
</ol>
<p>The motivation behind scenario 1 is to evaluate if it is possible to learn task-specific embeddings from a large but noisy dataset, but train and evaluate them on a less noisy location profiling task. For example, we could learn the embeddings on a large scale unlabeled dataset, and train and evaluate a location profiling model using a smaller labeled subset of that dataset. In this scenario, we consider a “best case” giving us an upper limit, because we use all data.</p>
<p>The motivation of scenario 2 is to evaluate the impact of noisy training data (i.e., approximations of the category associated with a location) on the embeddings learnt by the DeepCity framework and the location profiling model. Let’s take a look at the results obtained in both scenarios.</p>
</div></div>
<table><tbody>
<tr>
<th></th>
<th>Micro-F1</th>
<th>Macro-F1</th>
<th>Weighted-F1</th>
</tr>
<tr>
<td>Most Frequent</td>
<td>31.08%</td>
<td>5.93%</td>
<td>14.74%</td>
</tr>
<tr>
<td>Uniform</td>
<td>11.44%</td>
<td>9.94%</td>
<td>13.26%</td>
</tr>
<tr>
<td>Stratified</td>
<td>20.44%</td>
<td>12.19%</td>
<td>20.33%</td>
</tr>
<tr>
<td>Unbiased</td>
<td>34.20%</td>
<td>29.10%</td>
<td>34.83%</td>
</tr>
<tr>
<td>Biased</td>
<td>35.05%</td>
<td>26.76%</td>
<td>34.17%</td>
</tr>
<tr><td colspan="4"><b>Sentiance results: Scenario 1</b></td></tr>
</tbody></table>
<table><tbody>
<tr>
<th></th>
<th>Micro-F1</th>
<th>Macro-F1</th>
<th>Weighted-F1</th>
</tr>
<tr>
<td>Most Frequent</td>
<td>25.13%</td>
<td>5.02%</td>
<td>10.10%</td>
</tr>
<tr>
<td>Uniform</td>
<td>12.49%</td>
<td>10.83%</td>
<td>14.47%</td>
</tr>
<tr>
<td>Stratified</td>
<td>17.36%</td>
<td>9.84%</td>
<td>15.40%</td>
</tr>
<tr>
<td>Unbiased</td>
<td>21.68%</td>
<td>19.96%</td>
<td>21.21%</td>
</tr>
<tr>
<td>Biased</td>
<td>21.90%</td>
<td>19.05%</td>
<td>20.49%</td>
</tr>
<tr><td colspan="4"><b>Sentiance results: Scenario 2</b></td></tr>
</tbody></table>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p>The main conclusion from Scenario 1 and 2 is that biasing the random walks does not help when the biases are determined using noisy training data (labels). Let’s take a closer look at the different “views” used in the experiments.</p>
<p>In this section we will consider the following:</p>
<ul>
<li>What is the influence of the views on the location category distribution?</li>
<li>How much “conflicts” arise?</li>
<li>What is the influence on the bias values?</li>
</ul>
<p>Below is the distribution of categories considering different “views” on the data. We want to compare which type of places are visited at a certain location, versus what type of places is present at that location according to our venues database. There is quite some difference between the different views.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/fig_11.png" width="877" height="348"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 11:</b><b>  </b>Distribution of categories using different “views” on the data.</p>
<p>As mentioned before, because geohashes are used, it can happen that one geohash corresponds to one or more venues, possibly with different categories. The following graph gives us an idea of the “confusion” typically seen. In general, there is only a single venue category that is being actively checked-in to in a given geohash.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/fig_12_categories_per_geohash-2.png" width="672" height="240"></div>
</figure>
</div>
</div></div></div></div>
<div><div><div><div>
<div><div>
<p><b>Figure 12:</b><b>  </b>Number of distinct categories per geohash.</p>
<p>Finally, let’s take a closer look at the bias values used in the different experiments. The “approximate” views (checkins view and infrastructure view) are similar, but they are quite different from the biases used in the ground truth case, which may explain why the (biased) task-specific walks no longer provide many benefits.</p>
</div></div>
<div>
<figure><div><img src="https://www.sentiance.com/wp-content/uploads/2018/01/fig_13_biases.png" width="698" height="266"></div>
</figure>
</div>
<div><p><b>Figure 13:</b><b>  </b>Violin plots of the bias values obtained using the different views.</p></div>
<div><div>
<h2><span>Conclusion</span></h2>
<p>We have evaluated the location profiling capabilities from the DeepCity framework (<a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Pang, 2017</a>) and showed that it is capable of learning representations for location profiling tasks, that capture different types of information. Task-specific random walks can be used to guide the learning algorithm to learn similarities that are more relevant to the task, for example, category prediction.</p>
<p>We have shown how the framework can be extended to support unlabeled datasets. However, in such a use case we must be careful with using task-specific random walks based on possibly noisy labeled data. Experiments have indicated that in a scenario with noisy training data, task-specific random walks may have a negative effect on the performance.</p>
<p>We also evaluated the potential of the framework when labels are sparse, which would allow hybrid approaches, where the embeddings are trained on large unlabeled datasets, and specific models that depend on them are trained using a limited amount of labeled data by benefiting from the generalization properties of these type of embeddings.</p>
<p>We are currently integrating DeepCity location and/or user embeddings (together with other types of embeddings) in our existing models such as the venue mapping algorithm. You can test our latest innovations yourself by installing <a href="http://www.sentiance.com/demo/" target="_blank">Journeys</a> or getting in touch with our sales team.</p>
</div></div>
</div></div></div></div>
<div><div><div><div><div><div>
<h2><span>References</span></h2>
<ol>
<li><a rel="noopener" href="https://arxiv.org/abs/1301.3781" target="_blank">Tomas Mikolov, Kai Chen, Greg Corrado, Jeffrey Dean, <i>“Efficient Estimation of Word Representations in Vector Space”</i>, ICLR 2013 Workshop, 2013.</a></li>
<li><a rel="noopener" href="https://arxiv.org/abs/1310.4546" target="_blank">Tomas Mikolov, Ilya Sutskever, Kai Chen, Greg S. Corrado, Jeff Dean, <i>“Distributed Representations of Words and Phrases and their Compositionality”</i>, Advances in Neural Information Processing Systems 26 (NIPS 2013), 2013.</a></li>
<li><a rel="noopener" href="https://arxiv.org/abs/1403.6652" target="_blank">Bryan Perozzi, Rami Al-Rfou, Steven Skiena,<i> “DeepWalk: Online Learning of Social Representations”</i>, Proceedings of the 20th ACM SIGKDD international conference on Knowledge discovery and data mining (KDD ’14), 2014.</a></li>
<li><i>Dingqi Yang, Daqing Zhang, Vincent W. Zheng, Zhiyong Yu. Modeling User Activity Preference by Leveraging User Spatial Temporal Characteristics in LBSNs. IEEE Trans. on Systems, Man, and Cybernetics: Systems, (TSMC), 45(1), 129-142, 2015.</i></li>
<li><a rel="noopener" href="https://arxiv.org/abs/1610.03676" target="_blank">Jun Pang, Yang Zhang, <i>“DeepCity: A Feature Learning Framework for Mining Location Check-ins”</i>, Eleventh International AAAI Conference on Web and Social Media (ICWSM ’17), 2017.</a></li>
</ol>
</div></div></div></div></div></div>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via Sentiance https://ift.tt/2crG3Nu<br>
<br>
October 30, 2018 at 02:31PM
